fish-rust/src/function.rs

```diff
@@ -0,0 +1,711 @@
+        !has_explicit_func && !tombstoned
+    }
+}
+
+/// The big set of all functions.
+static FUNCTION_SET: Lazy<Mutex<FunctionSet>> = Lazy::new(|| {
+    Mutex::new(FunctionSet {
+        funcs: HashMap::new(),
+        autoload_tombstones: HashSet::new(),
+        autoloader: ffi::make_autoload_ffi(L!("fish_function_path").to_ffi()),
+    })
+});
+
+/// Necessary until autoloader has been ported to Rust.
+unsafe impl Send for FunctionSet {}
+
+/// Make sure that if the specified function is a dynamically loaded function, it has been fully
+/// loaded. Note this executes fish script code.
+fn load(name: &wstr, parser: &mut parser_t) -> bool {
+    parser.assert_can_execute();
+    let mut path_to_autoload: Option<WString> = None;
+    // Note we can't autoload while holding the funcset lock.
+    // Lock around a local region.
+    {
+        let mut funcset: std::sync::MutexGuard<FunctionSet> = FUNCTION_SET.lock().unwrap();
+        if funcset.allow_autoload(name) {
+            let path = funcset
+                .autoloader
+                .as_mut()
+                .unwrap()
+                .resolve_command_ffi(&name.to_ffi() /* Environment::globals() */)
+                .from_ffi();
+            if !path.is_empty() {
+                path_to_autoload = Some(path);
+            }
+        }
+    }
+
+    // Release the lock and perform any autoload, then reacquire the lock and clean up.
+    if let Some(path_to_autoload) = path_to_autoload.as_ref() {
+        // Crucially, the lock is acquired after perform_autoload().
+        ffi::perform_autoload_ffi(&path_to_autoload.to_ffi(), parser.pin());
+        FUNCTION_SET
+            .lock()
+            .unwrap()
+            .autoloader
+            .as_mut()
+            .unwrap()
+            .mark_autoload_finished(&name.to_ffi());
+    }
+    path_to_autoload.is_some()
+}
+
+/// Insert a list of all dynamically loaded functions into the specified list.
+fn autoload_names(names: &mut HashSet<WString>, get_hidden: bool) {
+    // TODO: justify this.
+    let vars = EnvStack::principal();
+    let Some(path_var) = vars.get_unless_empty(L!("fish_function_path")) else {
+        return;
+    };
+    let path_list = path_var.as_list();
+    for ndir_str in path_list {
+        let Ok(mut dir) = DirIter::new(ndir_str) else {
+            continue;
+        };
+        while let Some(entry) = dir.next() {
+            let Ok(entry) = entry else {
+                continue;
+            };
+            let func: &WString = &entry.name;
+            if !get_hidden && func.char_at(0) == '_' {
+                continue;
+            }
+            let suffix: Option<usize> = func.chars().rposition(|x| x == '.');
+            // We need a ".fish" *suffix*, it can't be the entire name.
+            if let Some(suffix) = suffix {
+                if suffix > 0 && entry.name.slice_from(suffix) == ".fish" {
+                    // Also ignore directories.
+                    if !entry.is_dir() {
```

Comment:
TODO: move the "entry is dir" check before/outside the slicing logic, since that's just a bitwise check.

Reply:
`is_dir()` may call `stat()` so we wish to defer it as long as possible.

Reply:
Sorry, I thought we had already stat’d the entry by this point.